ipa-core/src/secret_sharing/vector/transpose.rs

```diff
@@ -0,0 +1,800 @@
+//! Bit matrix transpose operations.
+//!
+//! These are used to convert data to and from vectorized representation.
+//!
+//! For example, if there is a 32-bit value associated with each record, the natural storage
+//! representation for those values is something like a `Vec<BA32>`, with the vector indexed by
+//! records.
+//!
+//! In vectorized code, we instead want to use something like `BitDecomposed<BA16>`, where each
+//! `BA16` holds the value of a particular bit for each of 16 records, and the `BitDecomposed`
+//! (which is just a wrapper around a `Vec`) is indexed by bits.
+//!
+//! To convert between these representations we need to transpose a 16x32 bit matrix into a 32x16
+//! bit matrix.
+//!
+//! This module stores bytes and bits are in little-endian order. Less significant bytes store data
+//! closer to the top or left of the matrix. Within each byte, the first (leftmost) column is in
+//! the least significant bit, and the last (rightmost) column is in the most significant bit.
+//!
+//! These implementations are somewhat optimized; it is certainly possible to optimize further, but
+//! that is only worthwhile if profiling indicates this is a significant contributor to our overall
+//! performance. Also, for some functions, the generated code is much better with `codegen-units =
+//! 1` than with the default codegen-units (unfortunately, `codegen-units = 1` significantly
+//! increases compile time). See [rust issue 47745](https://github.com/rust-lang/rust/issues/47745).
+//!
+//! Some possibilities for further optimization:
+//!  * Use codegen-units = 1 or figure out how to get comparable codegen without it.
+//!  * Avoid cost of zero-initializing the transpose destination.
+//!  * Use Rust's portable SIMD abstraction (not yet stable as of early 2024), or code directly
+//!    against platform SIMD intrinsics.
+//!
+//! For more ideas on optimizing bit matrix transposes in rust, see:
+//!  * <https://stackoverflow.com/a/77596340>
+//!  * <https://github.com/swenson/binary_matrix/tree/simd-transpose>
+
+// This rule throws false positives on "MxN".
+#![allow(clippy::doc_markdown)]
+
+use std::array;
+
+use crate::{
+    ff::{
+        boolean::Boolean,
+        boolean_array::{BA16, BA256, BA64},
+    },
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare, ReplicatedSecretSharing},
+        BitDecomposed, SharedValue, StdArray,
+    },
+};
+
+/// Trait for overwriting a value with the transpose of a source value.
+pub trait TransposeFrom<T> {
+    fn transpose_from(&mut self, src: T);
+
+    fn transposed_from(src: T) -> Self
```

Comment:
you don't seem to use it anywhere, are the plans to use it outside of this module?

Reply:
Yes, it is used in the vectorized share conversion protocol.

(I actually think it's pretty rare that we want to transpose on top of existing data vs. allocate new storage containing transposed data -- but a bunch of our types are not `Default`, which makes it hard to implement the `transposed_from` API in safe rust.)